File: 10.js
// graph ma'lumotlar tuzilmasi va breadth first search algoritm
let graph = {
  zokirkhon: ["ali", "vali", "tohir"],
  ali: ["aziza", "olim"],
  vali: ["botir", "ziyoda"],
  tohir: ["elonmusk", "mohir"],
  aziza: [],
  olim: [],
  botir: [],
  ziyoda: ["aziza"],
  elonmusk: [],
  mohir: [],
};


function search(graph, startNode, target) {
  let searchQueue = [];
  searchQueue.push(...graph[startNode]);
  let searched = [];

  while(searchQueue){
    //   console.log(searchQueue)
      let shifted = searchQueue.shift()
      if(!searched.includes(shifted)){
          if(shifted == target){
              console.log(`${target} ni topdik!`)
              console.log(searched)
              return true;
          } else {
              searchQueue.push(...graph[shifted])
              searched.push(shifted)
          }
      }
  }
  return false
}

console.log(search(graph, "zokirkhon", "mohir"))

File: 10.py
# graph ma'lumotlar tuzilmasi va breadth first search algoritm
from collections import deque;

graph = {
    'zokirkhon': ['ali', 'vali', 'tohir'],
    'ali': ['aziza', 'olim'],
    'vali': ['botir', 'ziyoda'],
    'tohir': ['elon musk', 'mohir'],
    'olim': [],
    'aziza': [],
    'botir': [],
    'ziyoda': ['aziza'],
    'elon musk': [],
    'mohir': []
}



def search(startNode,target):
    searchQueue = deque();
    searchQueue += graph[startNode]
    searched = set()
    
    while searchQueue:
        print(searchQueue)
        person = searchQueue.popleft();
        if person not in searched:
            if person == target:
                print(f"{target}ni topdik!")
                print(searched)
                return True;
            else:
                searchQueue += graph[person]
                searched.add(person)
    return False;


search('zokirkhon','botir')